### ISSUE TITLE
1942 - The Pacific Air War Gold - Sound/EMS issue

### ISSUE THREAD

**Issue by robrich82** [dosbox-staging/dosbox-staging#3920]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

0.82.0-rc1

### What Operating System are you using?

Windows 11

### Whats your question and how can we help?

I bought 1942 The Pacific Air War Gold off GOG. The configuration file provided didn't work in staging so I attempted to figure out why. The game runs if I remove the call to JEMMEX, but then the sound has issues per this thread https://www.vogons.org/viewtopic.php?t=41643

```[sdl]
fullscreen = true

[dosbox]
memsize = 32

[dos]
xms = false
ems = false
umb = false

[autoexec]
@ECHO OFF
CLS
C:
CD \MPS\1942
GOTO launcher

:launcher
CLS
ECHO [1;33m[42m╔════════════════════════════════════════════╗
ECHO ║ ------------------------------------------ ║
ECHO ║  1942: The Pacific Air War Launcher        ║
ECHO ║ ------------------------------------------ ║
ECHO ║  1) 1942: The Pacific Air War GOLD         ║
ECHO ║  2) Sound Settings                         ║
ECHO ║ ------------------------------------------ ║
ECHO ║  3) exit program                           ║
ECHO ║ ------------------------------------------ ║
ECHO ╚════════════════════════════════════════════╝[0m

CHOICE /c123 /s Which program do you want to run? [1-3]: /n 
if errorlevel 3 GOTO exit
if errorlevel 2 GOTO setup
if errorlevel 1 GOTO game

:game
CLS
JEMMEX.EXE FRAME=E000 LOAD
1942
GOTO exit

:setup
CLS
install.exe
GOTO launcher

:exit
EXIT
```

Running with this config gives the following output:
![Screenshot 2024-09-09 163022](https://github.com/user-attachments/assets/3eeb49cd-5328-40b4-b996-33c95f9fe7a5)

If I don't call JEMMEX and enable EMS in the configuration, then the sound doesn't play correctly.

With JEMMEX and the DOSBOX provided with GOG it works as expected.

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by johnnovak** [comment]
@robrich82 Try `memsize = 16` as per the linked Vogons thread. If that doesn't work, try 31 too.

**Comment by robrich82** [comment]
Both 16 and 31 give me the same error unfortunately.

**Comment by johnnovak** [comment]
> Both 16 and 31 give me the same error unfortunately.

Could you test it with earlier Staging version? Most likely this is a regression.

**Comment by MasterO2** [comment]
@robrich82, did you set `ems = true` in your batch file?  The batch file's contents currently has it set to **false**.

**Comment by johnnovak** [comment]
> @robrich82, did you set `ems = true` in your batch file? The batch file's contents currently has it set to **false**.

Why should he? The point here is using JEMMEX to get EMS memory instead of the built-in EMS support of DOSBox. That is why you have `ems = false` in the Vogons thread and the GOG config.

The game must use some non-emulated aspect of EMS memory, hence the need for a native driver.

**Comment by Grounded0** [comment]
1942: The Pacific Air War workaround added:

https://github.com/dosbox-staging/dosbox-staging/wiki/Game-issues#1942-the-pacific-air-war-including-gold

**Comment by johnnovak** [comment]
Nice one @Grounded0. This seems like a very unique and interesting game!

**Comment by Grounded0** [comment]
I'm sourcing my information from here:

- https://github.com/dosbox-staging/dosbox-staging/discussions/3768

Will add that fix then.

**Comment by interloper98** [comment]
@FeralChild64 , it seems like jemmex and the VMware interface are both using the same magic signature and API.

Maybe there more criteria the VMware IO read call can check before activiting? 

For example, the VMware interface is only used within Windows 3.x, so ideally it should only "come alive" if Windows 3.x is running. Is there a win 3.x signature we can look for?

Another alternative approach would be to acknowledge that this jemmex-vmware clash is truly unique, and therefore the VMware Read IO could check for the presence of jemmex and deactivate if it's present.

Another thought is we could default the VMware interface to false. But that's a bit heavy handed and users settings up the win3.1 driver would need to manually enable this conf setting. 

(Final option is to leave this as is, and inform jemmex users that they need to disable the VMware mouse interface)

**Comment by Grounded0** [comment]
Game bundled `JEMMEX.EXE` should now work so I can trim down that list.

**Comment by FeralChild64** [comment]
Windows tells us when it starts and ends, by calling int 0x2f - see how my WIP branch https://github.com/dosbox-staging/dosbox-staging/tree/fc/mouse-dos-windows-1 is using it. But I’m not sure if this is a good idea, as this interface is also used by the VBADOS mouse driver

Since JEMMEX is a free and open source tool (in fact, it is still being developed - see https://github.com/Baron-von-Riedesel/Jemm), I would start by checking what JEMMEX is trying to achieve and then make a decision what to do next.

I don’t have access to my PC at the moment - but maybe upgrading the JEMMEX to the latest revision would also help?

**Comment by Grounded0** [comment]
Someone should also check does updating `JEMMEX.EXE` to the latest release is a way to resolve the issue without having to tweak our stuff:

https://github.com/Baron-von-Riedesel/Jemm

**Comment by Grounded0** [comment]
My instructions currently state that integrated VMWare mouse driver should be off and Jemmex should be updated to v5.84. The original author states that by just upgrading Jemmex to v5.84 is enough to make the game work:

- https://github.com/dosbox-staging/dosbox-staging/discussions/3768#discussioncomment-9775091

**Comment by FeralChild64** [comment]
@Grounded0 I suggest to first investigate and test, and only then update the Wiki.

——

It seems JEMM is actually trying to detect VMWare - https://github.com/Baron-von-Riedesel/Jemm/blob/34f65c7d680c088a4d741f9baf07822c85333e39/src/INIT16.ASM#L2432
I don’t have more time today for further investigation, unfortunately.

**Comment by Grounded0** [comment]
I'll drop the first step (VMWare mouse interface off) for now. Others are welcome to test and edit Wiki based on evidence.

**Comment by robrich82** [comment]
FYI I was already using the latest release of JEMM (5.84) so that doesn't solve the issue by itself.

I changed the command to `JEMMEX.EXE D=128 FRAME=E000 NOINVLPG NORAM LOAD` and it still didn't solve the issue, I had to do this to make it work still.

```[mouse]
vmware_mouse = false```

**Comment by Grounded0** [comment]
Reverted back to previous state based on evidence gathered through discovery.

**Comment by interloper98** [comment]
Right on @Grounded0 and @robrich82 for clarifying that the latest release of JEMM (5.84) is still affected and doesn't solve this on its own.

**Comment by johnnovak** [comment]
> Another alternative approach would be to acknowledge that this jemmex-vmware clash is truly unique, and therefore the VMware Read IO could check for the presence of jemmex and deactivate if it's present.

Yeah, the ideal approach would be if JEMMEX and the VMware stuff could sort out conflicts automatically between themselves. And it seems JEMMEX is already doing something along those lines, so maybe @FeralChild64 will only need to adjust the VMware API emulation slightly.

> Another thought is we could default the VMware interface to false. But that's a bit heavy handed and users settings up the win3.1 driver would need to manually enable this conf setting.

That would be preferable IMO as the VMware mouse stuff is a niche feature only useful for running Win 3.x games in windowed mode. But you could argue this game needs very specific setup steps anyway, so _not_ having to disable the VMware mouse via config wouldn't improve things much. And we're not aware of anything else needing JEMMEX. So it's a toss up.

> But I’m not sure if this is a good idea, as this interface is also used by the VBADOS mouse driver.

As above; niche feature. IMO it's OK to require users to explicitly enable very rarely used niche features and this in definitely one of them.

But hopefully getting the VMware API autodetection right would solve this and then the config defaults can stay.

**Comment by FeralChild64** [comment]
It seems to me the JEMM only checks for VMWare during the initialization:

```
;--- VMware detection
;--- questionable, because Qemu "responds" similar to VMware

VMwareDetect proc c

	mov eax, 564D5868h	;magic number (="VMXh")
	mov cx, 000ah		;command number (000A=get VMware version)
	xor ebx,ebx 		;command specific parameter
	mov dx, 5658h		;VMware IO port (="VX")
	in eax,dx			;"returns" version number in EAX
	cmp ebx, 564D5868h	;and magic number in EBX (="VMXh")
	setz al
	mov ah,0
	ret

VMwareDetect endp
```

and:

```
	invoke VMwareDetect
	.if ax
		mov dword ptr SystemMemory+0E8h, 'VVVV'
		mov dword ptr SystemMemory+0ECh, 'XXXX'
		invoke cprintf, CStr("VMware detected", LF)
	.endif
```

In plain English: if VMWare interface is found, then assume certain memory layout and mark pages:
- `E8`, `E9`, `EA`, `EB` as _VMWARE allocated, but possibly re-usable via I= (0e800-0ebffh)_ (quote from JEMM source code)
- `EC`, `ED`, `EE`, `EF` as _EXCLUDE = forbidden from commandline_

Not good… I could probably check the current page name in our VMware interface implementation, and if it’s `JEMM386` or `JEMMEX`, pretend we are not there :D But it would only work in our emulated DOS environment, not when a real MS-DOS / PC-DOS / DR-DOS, etc., is booted.

**Comment by LiquidProj3ct** [comment]
I cannot add nothing else except, as @robrich82 said, with vmware_mouse = false I can play this game with DOSBox Staging. I don't understand what's the problem with vmware_mouse = false, the game works fine, I think. Anyway, I hope this can be solved in an elegant way!

**Comment by LiquidProj3ct** [comment]
The NOVMW has been added to Jemmex:

https://github.com/Baron-von-Riedesel/Jemm/releases/tag/v5.85

Now you can use: 
JEMM585\jemmex.exe d=128 frame=e000 NOVMW load 
Before executing the game, and it will work in DosBox Staging, without need vmware_mouse = false
